rpcs3/rpcs3qt/pad_led_settings_dialog.cpp

```diff
@@ -0,0 +1,94 @@
+﻿#include "pad_led_settings_dialog.h"
+#include "pad_settings_dialog.h"
+#include <QColorDialog>
+
+pad_led_settings_dialog::pad_led_settings_dialog(const int& colorR, const int& colorG, const int& colorB, const bool& led_low_battery_blink, const bool& led_battery_indicator, const int& led_battery_indicator_brightness, QDialog * parent)
+    : QDialog(parent)
+    , ui(new Ui::pad_led_settings_dialog)
+    , m_cR(colorR)
+    , m_cG(colorG)
+    , m_cB(colorB)
+    , m_low_battery_blink(led_low_battery_blink)
+    , m_battery_indicator(led_battery_indicator)
+    , m_battery_indicator_brightness(led_battery_indicator_brightness)
+{
+	ui->setupUi(this);
+	setFixedSize(size());
+
+	// Fill the form
+	ui->hs_indicator_brightness->setValue(led_battery_indicator_brightness);
+	ui->cb_led_blink->setChecked(led_low_battery_blink);
+	ui->cb_led_indicate->setChecked(led_battery_indicator);
+	switch_groupboxes(m_battery_indicator);
+
+	connect(ui->b_close, &QAbstractButton::clicked, this, &pad_led_settings_dialog::accept);
+
+	connect(this, &pad_led_settings_dialog::pass_led_settings, qobject_cast<pad_settings_dialog*>(parent), &pad_settings_dialog::apply_led_settings);
+
+	connect(ui->b_apply, &QAbstractButton::clicked, this, &pad_led_settings_dialog::pass_led_settings_signal);
+
+	// Open QColorDialog
+	connect(ui->b_colorpicker, &QPushButton::clicked, [=]()
+	{
+		QColor led_color(m_cR, m_cG, m_cB);
+		QColorDialog dlg(led_color, this);
+		dlg.setWindowTitle(tr("LED Color"));
+		if (dlg.exec() == QColorDialog::Accepted)
+		{
+			const QColor newColor = dlg.selectedColor();
+			m_cR = newColor.red();
+			m_cG = newColor.green();
+			m_cB = newColor.blue();
+		}
+	});
+
+	connect(ui->hs_indicator_brightness, &QAbstractSlider::valueChanged, this, &pad_led_settings_dialog::update_slider_label);
+
+	connect(ui->cb_led_indicate, &QCheckBox::toggled, this, &pad_led_settings_dialog::switch_groupboxes);
+}
+
+pad_led_settings_dialog::~pad_led_settings_dialog()
+{
+	delete ui;
+}
+
+// Draw an indicator for the current LED color
+void pad_led_settings_dialog::paintEvent(QPaintEvent* event)
```

Comment:
somehow the code as well as the outcome looks weird.
I guess there should be a more elegant way that doesn't rely on hardcoded numbers.

Also, I think it's better to use something like: SetLedColor instead of this paintEvent

Reply:
Redone slightly in 259cc17. I moved this code into a separate function and made it so the sample is redrawn only when the color values have changed. I also tidied it up and got rid of absolute values, the shape dimensions are now linked to the label.

What do you mean by outcome? The shape? I think it looks good enough for what it does, however the rounded corners are a bit blurred due to antialiasing. Take a look, this is with and without AA:

![G38Mwj](https://user-images.githubusercontent.com/7739280/72530334-31299700-3880-11ea-92e0-e4be3102feec.png)
![AjvEd3](https://user-images.githubusercontent.com/7739280/72530333-31299700-3880-11ea-8cb0-f98b7e45bb9f.png)

Which do you think is better?

Reply:
What do we need the paintEvent for now that you can just call redraw_color sample?

Reply:
Regarding the paintEvent:

https://doc.qt.io/qt-5/qpainter.html

> Warning: When the paintdevice is a widget, QPainter can only be used inside a paintEvent() function or in a function called by paintEvent().

Reply:
Fair enough. Just for my own interest: did you try without?

Reply:
I did switch to drawing on a pixmap at one point, so it's indeed irrelevant. Will remove paintEvent and instead call the function directly when needed.

Reply:
No. It can always be improved at a later point